### PR TITLE
[Java] Add putDoubleAscii/parseDoubleAscii methods.

### PR DESCRIPTION
**Pull request by vyazelenko**
This PR implements new `putDoubleAscii` and `parseDoubleAscii` methods similar to what is available for `int` and `long` values:
- `putDoubleAscii` - formats given number in a fixed format using the smallest number of the output digits possible.
  
  The implementation is based on the algorithm described in the [Adams U. Ryū: Fast Float-to-String Conversion. In: Proceedings of the 39th ACM SIGPLAN Conference on Programming Language Design and Implementation PLDI 2018, New York, NY, USA: Association for Computing Machinery; 2018. p. 270–282](https://doi.org/10.1145/3192366.3192369).

  #### Performance
  The performance of the `putDoubleAscii` is compared against `Double#toString(double)` and `DecimalFormat#format(double)`. 
  ```  Benchmark                                                                  (value)  Mode  Cnt     Score    Error  Units
  MutableDirectBufferPutDoubleAsciiBenchmark.decimalFormat                         0  avgt   30    52.178 ±  0.114  ns/op
  MutableDirectBufferPutDoubleAsciiBenchmark.decimalFormat                      1.25  avgt   30    85.599 ±  0.458  ns/op
  MutableDirectBufferPutDoubleAsciiBenchmark.decimalFormat                   1000000  avgt   30   108.251 ±  1.394  ns/op
  MutableDirectBufferPutDoubleAsciiBenchmark.decimalFormat   2.2250738585072009e-308  avgt   30   836.042 ±  2.136  ns/op
  MutableDirectBufferPutDoubleAsciiBenchmark.decimalFormat      7.317770170789331E15  avgt   30   204.989 ±  4.621  ns/op
  MutableDirectBufferPutDoubleAsciiBenchmark.decimalFormat                   0.15625  avgt   30   101.243 ±  2.255  ns/op
  MutableDirectBufferPutDoubleAsciiBenchmark.decimalFormat    -2.4028443684544056E18  avgt   30   230.239 ±  2.971  ns/op
  MutableDirectBufferPutDoubleAsciiBenchmark.decimalFormat     1.0905441441816094E30  avgt   30   701.177 ± 21.754  ns/op
  MutableDirectBufferPutDoubleAsciiBenchmark.decimalFormat                 -Infinity  avgt   30    27.578 ±  0.083  ns/op
  MutableDirectBufferPutDoubleAsciiBenchmark.decimalFormat    1.7976930114055267e308  avgt   30  2781.515 ± 21.115  ns/op
  MutableDirectBufferPutDoubleAsciiBenchmark.decimalFormat         random:[1e-3,1e7)  avgt   30   256.312 ±  3.391  ns/op
  MutableDirectBufferPutDoubleAsciiBenchmark.doubleToString                        0  avgt   30    12.146 ±  0.072  ns/op
  MutableDirectBufferPutDoubleAsciiBenchmark.doubleToString                     1.25  avgt   30    29.957 ±  0.039  ns/op
  MutableDirectBufferPutDoubleAsciiBenchmark.doubleToString                  1000000  avgt   30    28.383 ±  0.036  ns/op
  MutableDirectBufferPutDoubleAsciiBenchmark.doubleToString  2.2250738585072009e-308  avgt   30   747.232 ±  5.167  ns/op
  MutableDirectBufferPutDoubleAsciiBenchmark.doubleToString     7.317770170789331E15  avgt   30    44.453 ±  0.115  ns/op
  MutableDirectBufferPutDoubleAsciiBenchmark.doubleToString                  0.15625  avgt   30    32.621 ±  0.198  ns/op
  MutableDirectBufferPutDoubleAsciiBenchmark.doubleToString   -2.4028443684544056E18  avgt   30    53.387 ±  1.076  ns/op
  MutableDirectBufferPutDoubleAsciiBenchmark.doubleToString    1.0905441441816094E30  avgt   30   411.012 ±  0.776  ns/op
  MutableDirectBufferPutDoubleAsciiBenchmark.doubleToString                -Infinity  avgt   30     2.716 ±  0.003  ns/op
  MutableDirectBufferPutDoubleAsciiBenchmark.doubleToString   1.7976930114055267e308  avgt   30   893.603 ±  0.947  ns/op
  MutableDirectBufferPutDoubleAsciiBenchmark.doubleToString        random:[1e-3,1e7)  avgt   30    92.372 ±  0.076  ns/op
  MutableDirectBufferPutDoubleAsciiBenchmark.unsafeBuffer                          0  avgt   30     6.015 ±  0.032  ns/op
  MutableDirectBufferPutDoubleAsciiBenchmark.unsafeBuffer                       1.25  avgt   30    64.552 ±  0.153  ns/op
  MutableDirectBufferPutDoubleAsciiBenchmark.unsafeBuffer                    1000000  avgt   30    83.231 ±  0.438  ns/op
  MutableDirectBufferPutDoubleAsciiBenchmark.unsafeBuffer    2.2250738585072009e-308  avgt   30    49.618 ±  0.299  ns/op
  MutableDirectBufferPutDoubleAsciiBenchmark.unsafeBuffer       7.317770170789331E15  avgt   30    55.761 ±  1.214  ns/op
  MutableDirectBufferPutDoubleAsciiBenchmark.unsafeBuffer                    0.15625  avgt   30    76.952 ±  0.544  ns/op
  MutableDirectBufferPutDoubleAsciiBenchmark.unsafeBuffer     -2.4028443684544056E18  avgt   30    55.162 ±  0.078  ns/op
  MutableDirectBufferPutDoubleAsciiBenchmark.unsafeBuffer      1.0905441441816094E30  avgt   30    56.276 ±  0.905  ns/op
  MutableDirectBufferPutDoubleAsciiBenchmark.unsafeBuffer                  -Infinity  avgt   30     6.206 ±  0.007  ns/op
  MutableDirectBufferPutDoubleAsciiBenchmark.unsafeBuffer     1.7976930114055267e308  avgt   30    57.677 ±  0.636  ns/op
  MutableDirectBufferPutDoubleAsciiBenchmark.unsafeBuffer          random:[1e-3,1e7)  avgt   30    47.788 ±  0.062  ns/op
  ```
  As can be seen from the numbers above the implementation is faster than `DecimalFormat` in all tested cases and is about **5x faster** for the `[1e-3,1e7)` numbers range and up to **48x** for the `1.7976930114055267e308` case.

  As for the `Double#toString(double)` here we have a split where one of the implementations is faster then the other depending on the number being formatted. In particular `Double#toString(double)` handles small numbers such as `1.25` and `1000000` extremely well. On the other hand the `putDoubleAscii` is about **2x** faster for the `[1e-3,1e7)` numbers range and up to **15x** faster for the `2.2250738585072009e-308` case.

- `parseDoubleAscii` - parses ASCII-formatted string containing double value and supports both fixed format (e.g. `0.125`, `00023.23`, `.5`, `2.` etc.) as well as scientific format (e.g. `-92666518056446206563E3`, `1.79e15`, `567e-2`).
  
  The implementation is based on the algorithm described in the [Daniel Lemire, Number Parsing at a Gigabyte per Second, Software: Practice and Experience 51 (8), 2021. arXiv.2101.11408v3 [cs.DS] 24 Feb 2021](https://arxiv.org/pdf/2101.11408.pdf) paper.
  As a fallback the `Double#parseDouble` method is used. The fallback is extremely rare, e.g. in a test with `100'000'000` random numbers the fallback was called `40179` times or just **0.0004%** of the time.

  #### Performance
  The performance comparison is done against the `Double#parseDouble` method.
  ```  Benchmark                                                                                       (value)  Mode  Cnt    Score   Error  Units
  MutableDirectBufferParseDoubleAsciiBenchmark.parseDouble                                              0  avgt   30   15.750 ± 0.020  ns/op
  MutableDirectBufferParseDoubleAsciiBenchmark.parseDouble                                           1.25  avgt   30   25.464 ± 0.030  ns/op
  MutableDirectBufferParseDoubleAsciiBenchmark.parseDouble                                        1000000  avgt   30   27.711 ± 0.312  ns/op
  MutableDirectBufferParseDoubleAsciiBenchmark.parseDouble                        2.2250738585072009e-308  avgt   30  448.997 ± 4.579  ns/op
  MutableDirectBufferParseDoubleAsciiBenchmark.parseDouble                             7317770170789331.0  avgt   30   89.619 ± 0.147  ns/op
  MutableDirectBufferParseDoubleAsciiBenchmark.parseDouble   0.156250000000000000000000000000000000000000  avgt   30   77.980 ± 0.288  ns/op
  MutableDirectBufferParseDoubleAsciiBenchmark.parseDouble                         -2402844368454405395.2  avgt   30  150.390 ± 0.548  ns/op
  MutableDirectBufferParseDoubleAsciiBenchmark.parseDouble                1090544144181609348835077142190  avgt   30  265.266 ± 2.708  ns/op
  MutableDirectBufferParseDoubleAsciiBenchmark.parseDouble                                        -2e3000  avgt   30   29.429 ± 0.297  ns/op
  MutableDirectBufferParseDoubleAsciiBenchmark.parseDouble                         1.7976930114055267e308  avgt   30  327.824 ± 1.310  ns/op
  MutableDirectBufferParseDoubleAsciiBenchmark.unsafeBuffer                                             0  avgt   30    5.484 ± 0.004  ns/op
  MutableDirectBufferParseDoubleAsciiBenchmark.unsafeBuffer                                          1.25  avgt   30    8.441 ± 0.080  ns/op
  MutableDirectBufferParseDoubleAsciiBenchmark.unsafeBuffer                                       1000000  avgt   30   10.265 ± 0.059  ns/op
  MutableDirectBufferParseDoubleAsciiBenchmark.unsafeBuffer                       2.2250738585072009e-308  avgt   30   21.136 ± 0.103  ns/op
  MutableDirectBufferParseDoubleAsciiBenchmark.unsafeBuffer                            7317770170789331.0  avgt   30   19.800 ± 0.019  ns/op
  MutableDirectBufferParseDoubleAsciiBenchmark.unsafeBuffer  0.156250000000000000000000000000000000000000  avgt   30   52.223 ± 0.066  ns/op
  MutableDirectBufferParseDoubleAsciiBenchmark.unsafeBuffer                        -2402844368454405395.2  avgt   30   44.799 ± 0.216  ns/op
  MutableDirectBufferParseDoubleAsciiBenchmark.unsafeBuffer               1090544144181609348835077142190  avgt   30  398.914 ± 4.732  ns/op
  MutableDirectBufferParseDoubleAsciiBenchmark.unsafeBuffer                                       -2e3000  avgt   30   12.009 ± 0.034  ns/op
  MutableDirectBufferParseDoubleAsciiBenchmark.unsafeBuffer                        1.7976930114055267e308  avgt   30   21.452 ± 0.037  ns/op
  ```
  As can be seen from the above the `parseDoubleAscii` is faster (up to **22 times**) in all tested cases except when the fallback is executed (e.g. `1090544144181609348835077142190`).

### CODE REVIEW

File: agrona/src/main/java/org/agrona/AsciiEncoding.java [hunk 1]
```diff
@@ -379,6 +738,163 @@ public static int parseEightDigitsLittleEndian(final long bytes)
...
+     * Determines if the given value is multiple to the given power of five, i.e. if the {@code value} is divisible
+     * by {@code 5^power}.
+     *
+     * @param value to check.
+     * @param power of five factor.s
+     * @return {@code true} if value is divisible by the given power of five.
+     */
+    public static boolean isMultipleOfPowerOfFive(final long value, final int power)
+    {
+        // FIXME: Avoid modulo here
```
Comment (mjpt777) [inline agrona/src/main/java/org/agrona/AsciiEncoding.java:762]: Is there an easy solution to this?

Comment (vyazelenko) [reply agrona/src/main/java/org/agrona/AsciiEncoding.java:762]: Yes, you by doing what is outlined in this [Faster Remainder by Direct Computation: Applications to Compilers and Software Libraries](https://arxiv.org/abs/1902.01961) paper.

**Comment by plokhotnyuk** [comment]
Have you considered using [The Schubfach way to render doubles](https://drive.google.com/file/d/1luHhyQF9zKlM8yJ1nebU0OgVYhfC6CBN/view)? It is more efficient (especially for the short mantissa) and has a guarantee to serialize into the shortest decimal representation.

[JDK11+ compatible Java sources with the license](https://github.com/c4f7fcce9cb06515/Schubfach/blob/3c92d3c9b1fead540616c918cdfef432bca53dfa/todec/src/math/FloatToDecimal.java)

[JDK8+ compatible adoption for Scala](https://github.com/plokhotnyuk/jsoniter-scala/blob/45753aac4b84611b24e9de6407c388f45a1300ab/jsoniter-scala-core/jvm/src/main/scala/com/github/plokhotnyuk/jsoniter_scala/core/JsonWriter.scala#L1876)

**Comment by vyazelenko** [comment]
@plokhotnyuk Yes, I did and decided to go with the Ryū for first implementation.

**Comment by plokhotnyuk** [comment]
> @plokhotnyuk Yes, I did and decided to go with the Ryū for first implementation.

The jsoniter-scala library used it as a 1st implementation of the faster `float`/`double` serialization too ;)

[Here](https://github.com/plokhotnyuk/jsoniter-scala/blob/6e6bb9d7bed6de341ce0b781b403eb671d008468/jsoniter-scala-core/jvm/src/main/scala/com/github/plokhotnyuk/jsoniter_scala/core/JsonWriter.scala#L1777-L2262) is a link on the most optimized Ryū implementation before switching to Schubfach.

**Pull request closed without merging**